packages/flutter/lib/src/widgets/selectable_region.dart

```diff
@@ -445,18 +477,63 @@ class SelectableRegionState extends State<SelectableRegion> with TextSelectionDe
     );
   }
 
-  void _startNewMouseSelectionGesture(DragDownDetails details) {
-    widget.focusNode.requestFocus();
-    hideToolbar();
-    _clearSelection();
+  void _startNewMouseSelectionGesture(TapDragDownDetails details) {
+    switch (_getEffectiveConsecutiveTapCount(details.consecutiveTapCount)) {
+      case 1:
+        widget.focusNode.requestFocus();
+        hideToolbar();
+        _clearSelection();
+      case 2:
+        _selectWordAt(offset: details.globalPosition);
+    }
+  }
+
+  void _handleMouseDragStart(TapDragStartDetails details) {
+    switch (_getEffectiveConsecutiveTapCount(details.consecutiveTapCount)) {
+      case 1:
+        _selectStartTo(offset: details.globalPosition);
+    }
   }
 
-  void _handleMouseDragStart(DragStartDetails details) {
-    _selectStartTo(offset: details.globalPosition);
+  SelectionPoint? _initialDragStartSelectionPoint;
+  bool _shouldMoveEndEdge(TapDragUpdateDetails details) {
+    assert(_selectionDelegate.value.startSelectionPoint != null || _initialDragStartSelectionPoint != null);
+    _initialDragStartSelectionPoint = _initialDragStartSelectionPoint ?? _selectionDelegate.value.startSelectionPoint!;
+    final bool isDraggingForwardFromOrigin = details.localOffsetFromOrigin.dx > 0;
+    final bool isPositionAboveStartingBaseline = details.localPosition.dy < _initialDragStartSelectionPoint!.localPosition.dy;
+    final bool isPositionClampedToStartingLine = isPositionAboveStartingBaseline && details.localPosition.dy >= _initialDragStartSelectionPoint!.localPosition.dy - _initialDragStartSelectionPoint!.lineHeight;
+    final bool isPositionAtOrBelowStartingBaseline = details.localPosition.dy >= _initialDragStartSelectionPoint!.localPosition.dy;
+
+    if (isDraggingForwardFromOrigin && isPositionClampedToStartingLine) {
+      return true;
+    } else if (!isDraggingForwardFromOrigin && isPositionClampedToStartingLine) {
+      return false;
+    }
+
+    return isPositionAtOrBelowStartingBaseline;
   }
 
-  void _handleMouseDragUpdate(DragUpdateDetails details) {
-    _selectEndTo(offset: details.globalPosition, continuous: true);
+  void _handleMouseDragUpdate(TapDragUpdateDetails details) {
+    switch (_getEffectiveConsecutiveTapCount(details.consecutiveTapCount)) {
+      case 1:
+        _selectEndTo(offset: details.globalPosition, continuous: true);
+      case 2:
+        if (_shouldMoveEndEdge(details)) {
+          // Hold the start edge at the word located at the beginning of the drag.
+          _selectStartTo(offset: details.globalPosition - details.offsetFromOrigin, continuous: true, textGranularity: TextGranularity.word);
```

Comment:
I did this in order to keep the first selected word as the origin of the selection. Only moving the end edge will result in the beginning of the first selected word being the origin of the selection instead of the entire word.

For example `Hello world |how| are you`

Drag forwards to next word `Hello world |how are| you`.

Drag backwards to the first word `|Hello world how| are you`.

I'm thinking another way we can do this is swapping the start and end edges. What do you think?

Reply:
`_shouldMoveEndEdge` also seems to fail when a `ListView` is inside of a `SelectionArea`. This is because the `Offsets` given by the `TapDragUpdateDetails` are based on the current viewport and we don't have access to the scrollables offset to be able to take it into account. I haven't found a great way to access a `Scrollable`s offset since `ScrollableState.maybeOf(context)` only works if the `Scrollable` is a parent of `SelectionArea` and not if it is a child of `SelectionArea`. Curious to hear your thoughts on this, maybe I am missing something.

Reply:
This problem is fixed when re-implementing at the paragraph level.